### ISSUE TITLE
Validate cluster topology before creating

### ISSUE THREAD

**Issue by baznikin** [rook/rook#9419]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

Validate basic topology rules (no same name elsewhere else allowed) and refuse to create or update cluster with error.

**What is use case behind this feature:**

I run Kubernetes cluster on baremetal and virtualised machines. Since some VMs co-located on same hardware I mindlessly set up this topology:

```
kubectl label node node-1 topology.rook.io/chassis=node-1
kubectl label node node-2 topology.rook.io/chassis=node-2
kubectl label node node-3-worker topology.rook.io/chassis=node-3
kubectl label node node-3-master topology.rook.io/chassis=node-3
```

and few hours can't figure out why OSDs on node-1 and node-2 (baremetal servers) fails with:

```
7f9a0c7cf080 -1 osd.19 0 init unable to update_crush_location: (22) Invalid argument
7f9a0c7cf080 -1 osd.19 0 mon_cmd_maybe_osd_create fail: '(22) Invalid argument': (22) Invalid argument
```

**Environment**:

**Comment by travisn** [comment]
@baznikin Are you saying the fix was to not set the chassis name to a node name? It is certainly an unhelpful message coming from Ceph. I see that Rook could validate the topology labels don't match node names, or other conflicts across the topology.

**Comment by baznikin** [comment]
@travisn yes, you are right, my mistake was same name for different hierarchy tiers. Thanks, this feature will help someone in the future.

**Comment by JensErat** [comment]
I experienced similar issues, but with `topology.kubernetes.io/region` and `topology.kubernetes.io/zone` set to the same value. In the end, I only realized what's going on when I started to read source code and running the commands manually.

This is "pretty nasty to debug", especially to those new to Ceph. I'd expect region=zone not to be too uncommon in many setups. rook printing a warning/error message to the log (or even exposing this to the status?) would have been really helpful. Actually, I think rook should deny creating those OSDs, it's easy to discover and will definitely fail on ceph side.

This somewhat seems like a good first issue for jumping in with some code -- if you can give a coarse outline where to tackle the issue, I'd be ready to jump in with some code. I guess somwhere around `pkg/operator/ceph/cluster/osd/topology.go` (determining topology) or `pkg/operator/ceph/cluster/osd/osd.go` (those determine-location-functions)?

**Comment by travisn** [comment]
@JensErat Thanks for taking a look at this validation. I'm thinking the following steps would be necessary:
- Query the K8s API for all the nodes and their labels. 
- If there is a nodeAffinity for OSDs, filter the nodes accordingly
- Analyze the topology labels across all the nodes
- If different levels of the topology (zone, datacenter, rack, etc) have the same value as another level, return an error

It is a rather heavy-weight validation compared to how rare the condition is, but I can't think of a simpler way. 

The validation could be placed anywhere really since it's not specific, but perhaps it could be called from `pkg/operator/ceph/cluster/osd/osd.go` inside `Start()` before the OSDs are created.

**Comment by JensErat** [comment]
I have the feeling I might have catched some "more easy special case" with zone=region _on the same OSD_ -- validating this seems pretty trivial. Maybe it is even in the original poster's case (hostname=chassis in the same OSD)?

I assumed ceph's restriction is "a value must be unique among all topology keys for a given OSD". In this case, I would assume checking this during OSD "deployment" or when calculating the topology for an OSD is sufficient. I still fail at finding proper documentation on the Ceph side, though.

If the value needs to be unique across all keys on all OSDs, I guess holding a reverse index inside some kind of cluster representation should be fine: something like a `map[value]struct{key, []OSDsHavingThisValue}`. If you want to add another OSD and the key is different -- error out. This would enable a computational cheap evaluation without looping over all those objects at runtime. I'm not sure if this will result in an issue with changing orders on controller restarts (on the other hand, the situation and error handling would still be better than the situation described above, with unclear error message in the OSD logs).

**Comment by travisn** [comment]
@JensErat Ah, you're right we can simplify this check and catch the most common mistakes. Even if it does not catch all cases, a simple check at least for the topology of a single OSD can catch if there are any duplicates for that OSD. 

I'm thinking a check could be added to the [updateLocationWithNodeLabels()](https://github.com/rook/rook/blob/c5783a77cfeda17725c756f5e0e2cc1a7bd9b5a6/pkg/operator/ceph/cluster/osd/osd.go#L699-L714) method in the [for loop](https://github.com/rook/rook/blob/c5783a77cfeda17725c756f5e0e2cc1a7bd9b5a6/pkg/operator/ceph/cluster/osd/osd.go#L708) that iterates over the topology labels. This method would need to return an err if a duplicate is found, which should then make the failure more discoverable.

**Comment by xu-cheng** [comment]
I want to add this bug is quite nasty due to the fact that `topology.kubernetes.io/region` and `topology.kubernetes.io/zone` set to the same value is quite common.

Would it make sense to workaround this issue by ignoring `topology.kubernetes.io/region` if it has the same value as `topology.kubernetes.io/zone`?

**Comment by travisn** [comment]
@xu-cheng That's a good suggestion. While it won't catch all cases, if this is the common pain point, then we can quickly add that check. Are you interested in opening the PR?

**Comment by travisn** [comment]
This is resolved already by #10383

**Comment by chrisludwig** [comment]
This seems only partly solved. We are facing the same issues as the op in regards to the following labels.

- topology.rook.io/chassis
- topology.rook.io/rack
- topology.rook.io/room
- topology.rook.io/row

We ended up with equal values for these in some environments, which led to the same error.

```
2023-03-01T06:07:23.434+0000 7f0a6061cf40 -1 osd.0 0 mon_cmd_maybe_osd_create fail: '(22) Invalid argument': (22) Invalid argument
2023-03-01T06:07:23.434+0000 7f0a6061cf40 -1 osd.0 0 init unable to update_crush_location: (22) Invalid argument
```

Introducing validation for only `region` and zone `labels` therefore seems insufficient.
